src/NServiceBus.Core/Hosting/Helpers/AssemblyValidator.cs

```diff
@@ -0,0 +1,98 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.IO;
+    using System.Reflection.Metadata;
+    using System.Reflection.PortableExecutable;
+    using System.Security.Cryptography;
+
+    class AssemblyValidator
+    {
+        public (bool shouldLoad, string reason) ValidateAssemblyFile(string assemblyPath)
+        {
+            using (var stream = File.OpenRead(assemblyPath))
+            using (var file = new PEReader(stream))
+            {
+                var hasMetadata = false;
+
+                try
+                {
+                    hasMetadata = file.HasMetadata;
+                }
+                catch (BadImageFormatException) { }
+
+                if (!hasMetadata)
+                {
+                    return (false, "File is not a .NET assembly.");
+                }
+
+                var reader = file.GetMetadataReader();
+                var assemblyDefinition = reader.GetAssemblyDefinition();
+
+                if (!assemblyDefinition.PublicKey.IsNil)
+                {
+                    var publicKey = reader.GetBlobBytes(assemblyDefinition.PublicKey);
+                    var publicKeyToken = GetPublicKeyToken(publicKey);
+
+                    if (IsRuntimeAssembly(publicKeyToken))
+                    {
+                        return (false, "File is a .NET runtime assembly.");
+                    }
+                }
+
+                return (true, "File is a .NET assembly.");
+            }
+        }
+
+        public static bool IsRuntimeAssembly(byte[] publicKeyToken)
+        {
+            var lowerInvariant = BitConverter.ToString(publicKeyToken).Replace("-", string.Empty).ToLowerInvariant();
+
+            //System
+            if (lowerInvariant == "b77a5c561934e089")
+            {
+                return true;
+            }
+
+            //System.Core +  mscorlib
+            if (lowerInvariant == "7cec85d7bea7798e")
+            {
+                return true;
+            }
+
+            //Web
+            if (lowerInvariant == "b03f5f7f11d50a3a")
+            {
+                return true;
+            }
+
+            //patterns and practices
+            if (lowerInvariant == "31bf3856ad364e35")
+            {
+                return true;
+            }
+
+            if (lowerInvariant == "cc7b13ffcd2ddd51")
```

Comment:
As mentioned in the sync meeting, since it's not clear when this token is used, and we know the existing comments aren't always correct anyway, I'll change this method to be "known Microsoft tokens".